File: OpenKMAddOn/src/com/openkm/openoffice/logic/DocumentLogic.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package com.openkm.openoffice.logic;

import com.openkm.openoffice.bean.OKMDocumentBean;
import com.openkm.openoffice.util.FileUtil;
import com.openkm.openoffice.util.Util;
import com.openkm.ws.client.AccessDeniedException_Exception;
import com.openkm.ws.client.Document;
import com.openkm.ws.client.OKMAuth;
import com.openkm.ws.client.OKMAuthService;
import com.openkm.ws.client.OKMDocument;
import com.openkm.ws.client.OKMDocumentService;
import com.openkm.ws.client.RepositoryException_Exception;
import java.io.File;
import java.io.FileOutputStream;
import java.net.URL;
import java.util.logging.Level;
import java.util.logging.Logger;
import javax.xml.namespace.QName;
import javax.xml.ws.BindingProvider;

/**
 *
 * @author jllort
 */
public class DocumentLogic {

    private static QName AuthServiceName = new QName("http://endpoint.ws.openkm.git.es/", "OKMAuthService");
    private static QName DocumentServiceName = new QName("http://endpoint.ws.openkm.git.es/", "OKMDocumentService");

    public static void create(String host, String username, String password, OKMDocumentBean document) throws OKMException {
        String token = "";
        OKMAuthService authService = null;
        OKMDocumentService docService = null;
        OKMAuth okmAuth = null;
        OKMDocument okmDocument = null;
        Document doc = new Document();

        try {
            authService = new OKMAuthService(new URL(host + "/OKMAuth?wsdl"), AuthServiceName);
            docService = new OKMDocumentService(new URL(host + "/OKMDocument?wsdl"), DocumentServiceName);
            okmAuth = authService.getOKMAuthPort();
            okmDocument = docService.getOKMDocumentPort();
            BindingProvider bpAuth = (BindingProvider) okmAuth;
            BindingProvider bpDocument= (BindingProvider) okmDocument;
            bpAuth.getRequestContext().put(BindingProvider.ENDPOINT_ADDRESS_PROPERTY, host+"/OKMAuth");
            bpDocument.getRequestContext().put(BindingProvider.ENDPOINT_ADDRESS_PROPERTY, host+"/OKMDocument");

            token = okmAuth.login(username, password);

            doc.setPath(document.getPath());
            okmDocument.create(token, doc, FileUtil.readFile(document.getLocalFilename()));

            // Logout OpenKM
            okmAuth.logout(token);
            token = "";
        } catch (Exception ex) {
            if (!token.equals("")) {
                try {
                  // Logout OpenKM
                        okmAuth.logout(token);
                } catch (AccessDeniedException_Exception ex1) {
                } catch (RepositoryException_Exception ex1) {
                }
            }
            throw new OKMException(ex.getMessage() + "\nStackTrace\n" + ex.getStackTrace());
        } 
    }

    public static OKMDocumentBean chekckout(String host, String username, String password, Document doc, String directoryPath) throws OKMException {
        String token = "";
        OKMAuthService authService = null;
        OKMDocumentService docService = null;
        OKMAuth okmAuth = null;
        OKMDocument okmDocument = null;
        OKMDocumentBean okmDocumentBean = new OKMDocumentBean();

        try {
            authService = new OKMAuthService(new URL(host + "/OKMAuth?wsdl"), AuthServiceName);
            docService = new OKMDocumentService(new URL(host + "/OKMDocument?wsdl"), DocumentServiceName);
            okmAuth = authService.getOKMAuthPort();
            okmDocument = docService.getOKMDocumentPort();
            BindingProvider bpAuth = (BindingProvider) okmAuth;
            BindingProvider bpDocument= (BindingProvider) okmDocument;
            bpAuth.getRequestContext().put(BindingProvider.ENDPOINT_ADDRESS_PROPERTY, host+"/OKMAuth");
            bpDocument.getRequestContext().put(BindingProvider.ENDPOINT_ADDRESS_PROPERTY, host+"/OKMDocument");

            token = okmAuth.login(username, password);

            okmDocument.checkout(token, doc.getPath());
            byte[] bytedoc = okmDocument.getContent(token, doc.getPath(), false);

            String fileName = Util.getLocalFilenameWithoutCollisions(doc, directoryPath);
            File file = new File(fileName);
            file.createNewFile();

            FileOutputStream fos = new FileOutputStream(file);
            fos.write(bytedoc);
            fos.flush();
            fos.close();

            okmDocumentBean = Util.copy(doc);
            okmDocumentBean.setLocalFilename(fileName);

            // Logout OpenKM
            okmAuth.logout(token);
            token = "";
        } catch (Exception ex) {
            if (!token.equals("")) {
                try {
                    // Logout OpenKM
                    okmAuth.logout(token);
                } catch (AccessDeniedException_Exception ex1) {
                } catch (RepositoryException_Exception ex1) {
                }
            }
            throw new OKMException(ex.getMessage() + "\nStackTrace\n" + ex.getStackTrace());
        }
        return okmDocumentBean;
    }

    public static void cancelCheckout(String host, String username, String password, OKMDocumentBean document) throws OKMException {
        String token = "";
        OKMAuthService authService = null;
        OKMDocumentService docService = null;
        OKMAuth okmAuth = null;
        OKMDocument okmDocument = null;

        try {
            authService = new OKMAuthService(new URL(host + "/OKMAuth?wsdl"), AuthServiceName);
            docService = new OKMDocumentService(new URL(host + "/OKMDocument?wsdl"), DocumentServiceName);
            okmAuth = authService.getOKMAuthPort();
            okmDocument = docService.getOKMDocumentPort();
            BindingProvider bpAuth = (BindingProvider) okmAuth;
            BindingProvider bpDocument= (BindingProvider) okmDocument;
            bpAuth.getRequestContext().put(BindingProvider.ENDPOINT_ADDRESS_PROPERTY, host+"/OKMAuth");
            bpDocument.getRequestContext().put(BindingProvider.ENDPOINT_ADDRESS_PROPERTY, host+"/OKMDocument");

            token = okmAuth.login(username, password);

            okmDocument.cancelCheckout(token, document.getPath());

            // Logout OpenKM
            okmAuth.logout(token);
            token = "";
        } catch (Exception ex) {
            if (!token.equals("")) {
                try {
                    // Logout OpenKM
                    okmAuth.logout(token);
                } catch (AccessDeniedException_Exception ex1) {
                } catch (RepositoryException_Exception ex1) {
                }
            }
            throw new OKMException(ex.getMessage() + "\nStackTrace\n" + ex.getStackTrace());
        }
    }

    public static void checkin(String host, String username, String password, OKMDocumentBean document) throws OKMException {
        String token = "";
        OKMAuthService authService = null;
        OKMDocumentService docService = null;
        OKMAuth okmAuth = null;
        OKMDocument okmDocument = null;

        try {
            authService = new OKMAuthService(new URL(host + "/OKMAuth?wsdl"), AuthServiceName);
            docService = new OKMDocumentService(new URL(host + "/OKMDocument?wsdl"), DocumentServiceName);
            okmAuth = authService.getOKMAuthPort();
            okmDocument = docService.getOKMDocumentPort();
            BindingProvider bpAuth = (BindingProvider) okmAuth;
            BindingProvider bpDocument= (BindingProvider) okmDocument;
            bpAuth.getRequestContext().put(BindingProvider.ENDPOINT_ADDRESS_PROPERTY, host+"/OKMAuth");
            bpDocument.getRequestContext().put(BindingProvider.ENDPOINT_ADDRESS_PROPERTY, host+"/OKMDocument");

            token = okmAuth.login(username, password);

            okmDocument.setContent(token, document.getPath(), FileUtil.readFile(document.getLocalFilename()));
            okmDocument.checkin(token, document.getPath(),"");

            // Logout OpenKM
            okmAuth.logout(token);
            token = "";
        } catch (Exception ex) {
            if (!token.equals("")) {
                try {
                    // Logout OpenKM
                    okmAuth.logout(token);
                } catch (AccessDeniedException_Exception ex1) {
                } catch (RepositoryException_Exception ex1) {
                }
            }
            throw new OKMException(ex.getMessage() + "\nStackTrace\n" + ex.getStackTrace());
        }
    }
}

File: MSOpenKMCore/MSOpenKMCore/logic/DocumentLogic.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using MSOpenKMCore.ws;
using MSOpenKMCore.bean;
using System.IO;
using MSOpenKMCore.util;

namespace MSOpenKMCore.logic
{
    public class DocumentLogic
    {

        // check out document
        public static MSOpenKMCore.bean.OKMDocument checkoutDocument(document doc, String type, String host, String username, String password)
        {
            String token = "";
            OKMAuth authService = null;
            MSOpenKMCore.ws.OKMDocument documentService = null;

            try
            {
                // Init services
                SSL.init(host);
                authService = new OKMAuth(host);
                documentService = new MSOpenKMCore.ws.OKMDocument(host);

                // OpenKM authentication
                token = authService.login(username, password);

                // Create a new document
                MSOpenKMCore.bean.OKMDocument oKMDocument = Util.copy(doc, type);

                // Creates new document
                FileStream fileStream = new FileStream(oKMDocument.getLocalFilename(), FileMode.CreateNew);

                // Checking out document
                documentService.checkout(token, doc.path);
                byte[] byteDoc = documentService.getContent(token, doc.path, false);

                // Write the data to the file, byte by byte.
                for (int i = 0; i < byteDoc.Length; i++)
                {
                    fileStream.WriteByte(byteDoc[i]);
                }

                fileStream.Close();

                // Logout OpenKM
                authService.logout(token);
                token = "";

                return oKMDocument;
            }
            catch (Exception e)
            {
                if (!token.Equals(""))
                {
                    // Logout OpenKM
                    authService.logout(token);
                    token = "";
                }
                String errorMSG = "DocumentLogic (checkoutDocument)\n" + e.Message + "\n\n" + e.StackTrace;
                throw new OKMException(errorMSG);
            }
        }

        // cancel checkout document
        public static void cancelCheckout(MSOpenKMCore.bean.OKMDocument oKMDocument, String host, String username, String password)
        {
            String token = "";
            OKMAuth authService = null;
            MSOpenKMCore.ws.OKMDocument documentService = null;

            try
            {
                // Init services
                SSL.init(host);
                authService = new MSOpenKMCore.ws.OKMAuth(host);
                documentService = new MSOpenKMCore.ws.OKMDocument(host);

                // OpenKM authentication
                token = authService.login(username, password);

                // Cancelling checkout
                documentService.cancelCheckout(token, oKMDocument.getPath());

                // Logout OpenKM
                authService.logout(token);
                token = "";
            }
            catch (Exception e)
            {
                if (!token.Equals(""))
                {
                    // Logout OpenKM
                    authService.logout(token);
                    token = "";
                }
                String errorMSG = "DocumentLogic (cancelCheckout)\n" + e.Message + "\n\n" + e.StackTrace;
                throw new OKMException(errorMSG);
            }
        }

        // Checkin document
        public static void checkin(MSOpenKMCore.bean.OKMDocument oKMDocument, String host, String username, String password)
        {
            String token = "";
            MSOpenKMCore.ws.OKMAuth authService = null;
            MSOpenKMCore.ws.OKMDocument documentService = null;

            try
            {
                // Init services
                SSL.init(host);
                authService = new MSOpenKMCore.ws.OKMAuth(host);
                documentService = new MSOpenKMCore.ws.OKMDocument(host);

                // OpenKM authentication
                token = authService.login(username, password);

                // Checkin document
                //documentService.setContent(token, oKMDocument.getPath(), FileUtil.readFile(oKMDocument.getLocalFilename()));
                //documentService.checkin(token, oKMDocument.getPath(), "");

                // Logout OpenKM
                authService.logout(token);
                token = "";
            }
            catch (Exception e)
            {
                if (!token.Equals(""))
                {
                    // Logout OpenKM
                    authService.logout(token);
                    token = "";
                }
                String errorMSG = "DocumentLogic (checkin)\n" + e.Message + "\n\n" + e.StackTrace;
                throw new OKMException(errorMSG);
            }
        }

        // create document
        public static void create(String localFileName, String docPath, String host, String username, String password)
        {
            String token = "";
            OKMAuth authService = null;
            MSOpenKMCore.ws.OKMDocument documentService = null;

            try
            {
                // Init services
                SSL.init(host);
                authService = new OKMAuth(host);
                documentService = new MSOpenKMCore.ws.OKMDocument(host);

                // OpenKM authentication
                token = authService.login(username, password);

                // create document
                document doc = new document();
                doc.path = docPath;
                documentService.create(token, doc, FileUtil.readFile(localFileName));

                // Logout OpenKM
                authService.logout(token);
                token = "";
            }
            catch (Exception e)
            {
                if (!token.Equals(""))
                {
                    // Logout OpenKM
                    authService.logout(token);
                    token = "";
                }
                String errorMSG = "DocumentLogic (create)\n" + e.Message + "\n\n" + e.StackTrace;
                throw new OKMException(errorMSG);
            }
        }
    }
}
